### ISSUE TITLE
More lock issues -333

### ISSUE THREAD

**Issue by enkelmedia** [skttl/umbraco-fulltextsearch8#110]
Hi!

After upgrading Umbraco to 12.3+ we started to see lock issues, at first we thought it was related to something in Core and it might be that they have made some changes that impacts this.

However, after disabling the Alpha-version of FullTextSearch the lock issues went away =/  Re-enabled FullTextSearch and lock-issues is back. We do not have any other event-handlers so I'm still not 100% sure that the problem is in the package-alpha. I'm trying right now by making sure that all scopes are explicitly completed and not rely on "autoComplete=true".

It looks like this might have a positive impact - if that's the case I'm happy to share a PR soon.

```
Server error: Contact administrator, see log for full details.
Failed to acquire write lock for id: -333.

https://our.umbraco.com/forum/using-umbraco-and-getting-started/101745-lock-request-time-out-period-exceeded

Microsoft.Data.SqlClient.SqlException (0x80131904): Lock request time out period exceeded.
The statement has been terminated.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean isAsync, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String method)
   at Microsoft.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String methodName)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func)
   at NPoco.Database.ExecuteNonQueryHelper(DbCommand cmd)
   at NPoco.Database.Execute(String sql, CommandType commandType, Object[] args)
ClientConnectionId:3ded19d8-ec90-4a49-ba53-3ea31f9702f6
Error Number:1222,State:51,Class:16
```

**Comment by skttl** [comment]
Thanks! I only have a few projects running the alpha and haven't hit those issues. Your testing is very much appreciated. I'm going to hold back the final release a bit more (also want to do that config schema thing, Corné wrote about :))

**Comment by enkelmedia** [comment]
Sounds fair!

We had the issue on 12.3.1 and 12.3.2,

After upgrading to 12.3.3 and updating all the scopes in FullTextSearch to be explicitly closed the problem seems to be gone. The problem is not that I'm not 100% sure if the changes in our FullTextSearch was the solution or if this fix might have hade an impact https://github.com/umbraco/Umbraco-CMS/issues/15207 maybe both.

I'm going to try to rollback to 12.3.2 and see if we can replicate the issue and isolate the solution.

I'll keep you posted =D

**Comment by enkelmedia** [comment]
Hi again!

I must admit a rookie-misstake =D I was 100% sure that the project was on `alpha-004` so I did not even look close. After 4 hours of debugging, updating the .NET runtime, and so on I noticed that it was actually on `alpha-001`. 

The problem with locks probably originates from before this PR was added https://github.com/skttl/umbraco-fulltextsearch8/pull/102/files where we added `ConfigureAwait()` and the changes in the execution order of the NotificationHandlers. It's very likely that the implementation where they would run simultaneously created race conditions from time to time.

Anyway, after updating to `alpha-004` I can not replicate the problem anymore on the latest Umbraco-version (right now `12.3.3`).

I'll go ahead and close this one, sorry for raising unneeded concerns.

Today's words of wisdom: Assumptions is the mother of all fuckups.
